### PR TITLE
Implement asset-listing

### PR DESCRIPTION
**Pull request by dpfaffenbauer**
This is sort of a POC, there are some things I am sure are not 100% right yet.

### CODE REVIEW

File: src/GraphQL/Resolver/Element.php [hunk 1]
```diff
@@ -136,6 +148,165 @@ public function resolveSiblings($value = null, $args = [], $context, ResolveInfo
...
+//
+//        if ($sqlListCondition) {
+//            $conditionParts[] = '(' . $sqlListCondition . ')';
+//        }
+
+        // check permissions
+        $db = Db::get();
+        $tableName = $this->getTypeTableName();
+        $workspacesTableName = 'plugin_datahub_workspaces_' . $this->elementType;
+        $conditionParts[] = ' (
```
Comment (dpfaffenbauer) [inline src/GraphQL/Resolver/Element.php:239]: this query doesn't work properly, not sure why, I copied it from objects

File: src/GraphQL/Resolver/Element.php [hunk 2]
```diff
@@ -136,6 +148,165 @@ public function resolveSiblings($value = null, $args = [], $context, ResolveInfo
...
+            $filterCondition = Helper::buildSqlCondition($tableName, $filter);
+            $conditionParts[] = $filterCondition;
+        }
+
+        if ($conditionParts) {
+            $condition = implode(' AND ', $conditionParts);
+            $objectList->setCondition($condition);
+        }
+
+//        $event =  new ListingEvent(
```
Comment (dpfaffenbauer) [inline src/GraphQL/Resolver/Element.php:271]: Should we use the same event or introduce a new one?

File: src/GraphQL/Resolver/Element.php [hunk 3]
```diff
@@ -136,6 +148,165 @@ public function resolveSiblings($value = null, $args = [], $context, ResolveInfo
...
+        }
+
+        $modelFactory = $this->getGraphQlService()->getModelFactory();
+        $listClass = $this->getListType();
+
+        /** @var AbstractListing $objectList */
+        $objectList = $modelFactory->build($listClass);
+        $conditionParts = [];
+        if (isset($args['ids'])) {
+            $conditionParts[] = '(o_id IN (' . $args['ids'] . '))';
```
Comment (weisswurstkanone) [inline src/GraphQL/Resolver/Element.php:202]: there are no o_ids for assets

File: src/GraphQL/Resolver/Element.php (hunk 1 above)
Comment (weisswurstkanone) [reply src/GraphQL/Resolver/Element.php:239]: what is the query and what is the error ?

File: src/GraphQL/Resolver/Element.php (hunk 2 above)
Comment (weisswurstkanone) [reply src/GraphQL/Resolver/Element.php:271]: current one is generic enough I guess !?

File: src/GraphQL/Resolver/Element.php (hunk 1 above)
Comment (dpfaffenbauer) [reply src/GraphQL/Resolver/Element.php]: fixed it

File: src/GraphQL/Resolver/Element.php (hunk 2 above)
Comment (dpfaffenbauer) [reply src/GraphQL/Resolver/Element.php]: done

**Comment by dpfaffenbauer** [comment]
changed it a bit: introduced a new resolver for better distinction of asset and document

File: src/GraphQL/Resolver/AssetListing.php [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        // sorting
+        if (!empty($args['sortBy'])) {
+            $objectList->setOrderKey($args['sortBy']);
+            if (!empty($args['sortOrder'])) {
+                $objectList->setOrder($args['sortOrder']);
+            }
+        }
+
+        // Include unpublished
+//        if (isset($args['published']) && $args['published'] === false) {
```
Comment (weisswurstkanone) [inline src/GraphQL/Resolver/AssetListing.php:113]: this can be removed , no unpublished assets

File: src/GraphQL/Resolver/AssetListing.php [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+//            $objectList->setUnpublished(true);
+//        }
+
+        /** @var $configuration Configuration */
+        $configuration = $context['configuration'];
+//        $sqlListCondition = $configuration->getSql();
+//
+//        if ($sqlListCondition) {
```
Comment (weisswurstkanone) [inline src/GraphQL/Resolver/AssetListing.php:121]: this as well

File: src/Controller/ConfigController.php [hunk 6]
```diff
@@ -341,7 +341,7 @@ public function getAction(Request $request, Service $graphQlService): JsonRespon
             $config['schema']['specialEntities'] = [];
         }
 
-        $specialSettings = ["document", "document_folder", "asset", "asset_folder", "object_folder"];
+        $specialSettings = ["document", "document_folder", "asset", "asset_folder", "asset_listing", "object_folder"];
```
Comment (weisswurstkanone) [inline src/Controller/ConfigController.php:344]: translation key missing. probably something like "plugin_pimcore_datahub_graphql_special_asset_listing"

Comment (weisswurstkanone) [reply src/Controller/ConfigController.php:344]: besides that I think we should disable the configuration option other than "read"

File: src/GraphQL/Resolver/AssetListing.php [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+     * @param array $args
+     * @param $context
+     * @param ResolveInfo|null $resolveInfo
+     * @return array
+     * @throws \Exception
+     */
+    protected function extractSingleElement($element, $args, $context, $resolveInfo)
+    {
+        // Check Workspace permissions
+        if (!WorkspaceHelper::isAllowed($element, $context['configuration'], 'read')) {
```
Comment (weisswurstkanone) [inline src/GraphQL/Resolver/AssetListing.php:231]: let's see which one gets merged first :-)  https://github.com/pimcore/data-hub/pull/233

File: src/GraphQL/Resolver/AssetListing.php (hunk 4 above)
Comment (dpfaffenbauer) [reply src/GraphQL/Resolver/AssetListing.php]: done

File: src/GraphQL/Resolver/AssetListing.php (hunk 5 above)
Comment (dpfaffenbauer) [reply src/GraphQL/Resolver/AssetListing.php]: done

File: src/Controller/ConfigController.php (hunk 6 above)
Comment (dpfaffenbauer) [reply src/Controller/ConfigController.php:344]: translation key: done
I think we should keep this for now, I am planning to create a separate PR to allow to extend that from outside DataHub, I can refactor that with that then. WDYT?

File: src/GraphQL/Resolver/AssetListing.php (hunk 7 above)
Comment (dpfaffenbauer) [reply src/GraphQL/Resolver/AssetListing.php:221]: MINE :P

**Comment by weisswurstkanone** [comment]
Tried the following query:

```
{
  getAssetListing(defaultLanguage: "de") {
    edges {
      node {
        __typename
      }
    }
  }
}
```

but get

Attempted to call an undefined method named &quot;resolveEdges&quot; of class &quot;Pimcore\\Bundle\\DataHubBundle\\GraphQL\\Resolver\\Element&quot;.

**Comment by weisswurstkanone** [comment]
btw, can you mention this in the docs please ? maybe add a sample ...

**Comment by dpfaffenbauer** [comment]
error is fixed
docs are added

File: src/Controller/ConfigController.php (hunk 6 above)
Comment (weisswurstkanone) [reply src/Controller/ConfigController.php:344]: Sounds good!

**Comment by weisswurstkanone** [comment]
thx a lot!

**Pull request merged**